src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java

```diff
@@ -175,11 +179,34 @@ public ListenableFuture<Void> uploadBlobAsync(
         MoreExecutors.directExecutor());
   }
 
-  private static String buildUploadResourceName(
+  private boolean isOldStyleDigestFunction() {
+    // Old-style digest functions (SHA256, etc) are distinguishable by the length
+    // of their hash alone and do not require extra specification, but newer
+    // digest functions (which may have the same length hashes as the older
+    // functions!) must be explicitly specified in the upload resource name.
+    return digestFunction.getNumber() <= 7;
+  }
+
+  private String buildUploadResourceName(
```

Comment:
You may also want to change `GrpcCacheClient#getResourceName`.

Reply:
> You may also want to change `GrpcCacheClient#getResourceName`, maybe in another PR.

Yeah it's a bit cyclical; I'm happy to add support for it here so that download and upload are both kept in sync -- which I've done.

Adding another test for the download path will simple once there is a DigestHashFunction for BLAKE3. Will send that in a followup after adding the hasher.